File: src/claudia/javascript/features/steps/payments.js
const {BeforeAll, AfterAll, Given, When, Then, Before, After} = require('@cucumber/cucumber')
const xrpl = require("xrpl");
const {AccountSetAsfFlags} = require("xrpl");
const {} = require('../context.js');
let context = require('../context.js');
const assert = require('assert');
const ObjFactory = require("../lib/ObjFactory");
const {ObjType} = require("../lib/ObjFactory");
const {
    verify_account_balance,
    verify_payment_account_objects,
    ResponseCode,
    get_account_balance,
    verify_test_status,
    verify_test_error_message,
} = require("./common");

When('we send a payment of {string} XRP from Alice to Bob', {timeout: 180000}, async (transferAmount) => {
    try {

        context.transferAmount = ((transferAmount == ("10000000000000") || transferAmount.toString().includes(".")) ? transferAmount : xrpl.xrpToDrops(transferAmount))
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP from Alice to self', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress, "Amount": context.transferAmount, "Destination": context.sourceAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we only submit a payment of {string} XRP from Alice to Bob', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.payment_response = await context.client.submit(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment with amount greater than balance from Alice to Bob', {timeout: 180000}, async () => {
    try {
        context.transferAmount = await context.client.getXrpBalance(context.sourceAddress) + xrpl.xrpToDrops(1)
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP from Alice to Invalid', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress, "Amount": context.transferAmount, "Destination": context.invalidAddress,
        }));

        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.payment_response = await context.client.submit(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP to Bob without providing source information', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": "", "Amount": context.transferAmount, "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.payment_response = await context.client.submit(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP from Alice without providing destination information', {timeout: 180000}, async (transferAmount) => {
    //TODO: Revert this after https://github.com/XRPLF/xrpl.js/issues/2517 is fixed.
    return
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": "",
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.payment_response = await context.client.submit(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} USD from Alice to Bob', {timeout: 180000}, async (transferAmountInUSD) => {
    try {
        context.transferAmount = {
            currency: "USD", value: transferAmountInUSD, issuer: "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn"
        }
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP with destination tag {string} info from Alice to Bob', {timeout: 180000}, async (transferAmount, destinationTag) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
            "DestinationTag": parseInt(destinationTag)
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP with invoice id {string} from Alice to Bob', {timeout: 180000}, async (transferAmount, invoiceID) => {
    try {
        context.transferAmount = transferAmount
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
            "InvoiceID": invoiceID
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP from Alice as a preauth account to Bob', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        const autofilled_account_set_tx = await context.client.autofill({
            TransactionType: 'AccountSet',
            Account: context.destinationAddress,
            SetFlag: AccountSetAsfFlags.asfDepositAuth
        })
        let signed_autofilled_account_set_tx = context.destinationWallet.sign(autofilled_account_set_tx).tx_blob;
        await context.client.submitAndWait(signed_autofilled_account_set_tx);
        const account_info_request = {
            command: 'account_info', account: context.destinationAddress,
        }
        const last_sequence = (await context.client.request(account_info_request)).result.account_data.Sequence;
        const autofilled_deposit_preauth_tx = await context.client.autofill({
            TransactionType: "DepositPreauth",
            Account: context.destinationAddress,
            Sequence: last_sequence,
            Authorize: context.sourceAddress,
        })
        let signed_autofilled_deposit_preauth_tx = context.destinationWallet.sign(autofilled_deposit_preauth_tx).tx_blob;
        await context.client.submitAndWait(signed_autofilled_deposit_preauth_tx);
        context.destinationBalance = await context.client.getXrpBalance(context.destinationAddress)
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

When('we send a payment of {string} XRP from Alice as a non-preauth account to Bob', {timeout: 180000}, async (transferAmount) => {
    try {
        context.transferAmount = transferAmount
        const autofilled_account_set_tx = await context.client.autofill({
            TransactionType: 'AccountSet',
            Account: context.destinationAddress,
            SetFlag: AccountSetAsfFlags.asfDepositAuth
        })
        await context.client.submitAndWait(autofilled_account_set_tx, {
            wallet: context.destinationWallet,
        });
        context.destinationBalance = await context.client.getXrpBalance(context.destinationAddress)
        context.prepared = await context.client.autofill(ObjFactory.getObj(ObjFactory.ObjType.Payment, {
            "Account": context.sourceAddress,
            "Amount": context.transferAmount,
            "Destination": context.destinationAddress,
        }));
        context.signed_payment_tx = context.sourceWallet.sign(context.prepared).tx_blob;
        context.response = await context.client.submitAndWait(context.signed_payment_tx);
    } catch (error) {
        context.exception = error
        context.testStatus = "failed"
    }
});

Then('the balances after a valid payment are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - (xrpl.dropsToXrp(context.minDrop)) - xrpl.dropsToXrp(context.transferAmount)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance) + parseFloat(xrpl.dropsToXrp(context.transferAmount))));
});

Then('the balances are correct in submit only mode', {timeout: 180000}, async () => {
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the self-payment should fail', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /temREDUNDANT/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
});

Then('the payment with amount greater than balance should fail', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response, ResponseCode.unfunded_payment)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - xrpl.dropsToXrp(context.minDrop)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the payment with invalid destination should fail', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /Error: Non-base58 character/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
});

Then('the payment with zero amount should fail', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /temBAD_AMOUNT/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the payment with decimal amount should fail', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /Error: 10.5 is an illegal amount/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the payment without source information should fail', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /Missing field 'account'/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the payment without destination information should fail', {timeout: 180000}, async () => {
    // TODO: Fix after https://github.com/XRPLF/xrpl.js/issues/2517 is fixed.
    return
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /Serialized transaction does not match original txJSON/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
});

Then('the payment of 10 million XRP should fail', {timeout: 180000}, async () => {
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - (xrpl.dropsToXrp(context.minDrop))));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the balances after a non-xrp payment are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response, ResponseCode.path_dry)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - (xrpl.dropsToXrp(context.minDrop))));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the balances after a payment with destination tag info are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - (xrpl.dropsToXrp(context.minDrop)) - xrpl.dropsToXrp(context.transferAmount)).toFixed(6));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance) + parseFloat(xrpl.dropsToXrp(context.transferAmount))).toFixed(6));
});

Then('the balances after a payment with invoice id info are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - xrpl.dropsToXrp(context.transferAmount) - xrpl.dropsToXrp(context.minDrop)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance) + parseFloat(xrpl.dropsToXrp(context.transferAmount))));
});

Then('the balances after a payment with invalid invoice id info are correct', {timeout: 180000}, async () => {
    verify_test_status(context.testStatus, "failed")
    verify_test_error_message(context.exception.toString(), /Error: Invalid Hash length 1/, "Incorrect error message received.")
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

Then('the balances after a payment from preauth account are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - xrpl.dropsToXrp(context.transferAmount) - xrpl.dropsToXrp(context.minDrop)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance) + parseFloat(xrpl.dropsToXrp(context.transferAmount))));
});

Then('the balances after a payment from non-preauth account are correct', {timeout: 180000}, async () => {
    await verify_payment_account_objects(context.client, context.response, ResponseCode.no_permission)
    verify_account_balance(await get_account_balance(context, context.sourceAddress), (parseFloat(context.sourceBalance) - xrpl.dropsToXrp(context.minDrop)));
    verify_account_balance(await get_account_balance(context, context.destinationAddress), (parseFloat(context.destinationBalance)));
});

File: src/claudia/python/features/steps/payments.py
from behave import *
from claudia.python.lib.framework.common import *
from claudia.python.lib.framework.object_factory import ObjFactory, ObjType
from xrpl.account import get_balance
from xrpl.models import IssuedCurrencyAmount, DepositPreauth, AccountSet, AccountSetAsfFlag
from xrpl.models.requests import Tx
from xrpl.transaction import submit


@when('we send a payment of "{transfer_amount}" XRP from Alice to Bob')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = float(transfer_amount) if '.' in transfer_amount else int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )
        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP from Alice to self')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.sourceAddress,
            amount=str(context.transferAmount)
        )
        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we only submit a payment of "{transfer_amount}" XRP from Alice to Bob')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )
        context.signed_payment_tx = submit(context.payment, context.sourceWallet)
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment with amount greater than balance from Alice to Bob')
def step_impl(context):
    try:
        context.transferAmount = get_balance(context.sourceAddress, context.client) + 1
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP from Alice to Invalid')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.invalidAddress,
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP to Bob without providing source information')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )
        context.payment_response = submit_and_wait(context.payment, context.client, "")
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP from Alice without providing destination information')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount_in_usd}" USD from Alice to Bob')
def step_impl(context, transfer_amount_in_usd):
    try:
        context.transferAmount = IssuedCurrencyAmount(
            currency="USD",
            value=str(transfer_amount_in_usd),
            issuer="rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn"
        )

        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=context.transferAmount
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP with destination tag "{destination_tag}" info from Alice to Bob')
def step_impl(context, transfer_amount, destination_tag):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            destination_tag=int(destination_tag),
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP with invoice id "{invoice_id}" from Alice to Bob')
def step_impl(context, transfer_amount, invoice_id):
    try:
        context.transferAmount = int(transfer_amount)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            invoice_id=invoice_id,
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP from Alice as a preauth account to Bob')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)

        account_set_tx = AccountSet(
            account=context.destinationAddress,
            set_flag=AccountSetAsfFlag.ASF_DEPOSIT_AUTH,
        )

        submit_and_wait(account_set_tx, context.client, context.destinationWallet)

        deposit_pre_auth_tx = DepositPreauth(
            account=context.destinationAddress,
            sequence=get_account_sequence(context.client, context.destinationAddress),
            authorize=context.sourceAddress,
            fee=str(context.minDrop)
        )
        submit_and_wait(deposit_pre_auth_tx, context.client, context.destinationWallet)
        context.destinationBalance = get_balance(context.destinationAddress, context.client)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )

        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@when('we send a payment of "{transfer_amount}" XRP from Alice as a non-preauth account to Bob')
def step_impl(context, transfer_amount):
    try:
        context.transferAmount = int(transfer_amount)
        account_set_tx = AccountSet(
            account=context.destinationAddress,
            set_flag=AccountSetAsfFlag.ASF_DEPOSIT_AUTH,
        )

        submit_and_wait(account_set_tx, context.client, context.destinationWallet)

        context.destinationBalance = get_balance(context.destinationAddress, context.client)
        context.payment = ObjFactory.getObj(
            ObjType.payment,
            account=context.sourceAddress,
            destination=context.destinationAddress,
            amount=str(context.transferAmount)
        )
        context.payment_response = submit_and_wait(context.payment, context.client, context.sourceWallet)
        context.response = context.client.request(Tx(transaction=context.payment_response.result["hash"]))
    except Exception as e:
        context.exception = e
        context.test_status = "failed"


@then('the balances after a valid payment are correct')
def step_impl(context):
    verify_payment_account_objects(context, context.response)
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.transferAmount - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress), (
            context.destinationBalance + context.transferAmount))


@then('the balances are correct in submit only mode')
def step_impl(context):
    verify_account_balance(get_account_balance(context, context.sourceAddress), (context.sourceBalance))
    verify_account_balance(get_account_balance(context, context.destinationAddress), (context.destinationBalance))


@then('the self-payment should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception),
                              "An XRP payment transaction cannot have the same sender and destination.")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)


@then('the payment with amount greater than balance should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception),
                              "Transaction failed: tecUNFUNDED_PAYMENT")

    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the payment with invalid destination should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception),
                              "Invalid value to construct an AccountID: expected valid classic address or "
                              "X-Address, received str.")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)


@then('the payment with zero amount should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Can only send positive amounts.")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the payment with decimal amount should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Error processing Amount: 10.5 is an invalid XRP amount.")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the payment without source information should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Wallet must be provided when submitting an unsigned transaction")
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the payment without destination information should fail')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Destination not specified.")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)


@then('the payment of 10 million XRP should fail')
def step_impl(context):
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the balances after a non-xrp payment are correct')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Transaction failed: tecPATH_DRY")
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the balances after a payment with destination tag info are correct')
def step_impl(context):
    verify_payment_account_objects(context, context.response)
    verify_account_balance(get_account_balance(context, context.sourceAddress), (
            context.sourceBalance - context.transferAmount - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress),
                           (context.destinationBalance + context.transferAmount))


@then('the balances after a payment with invoice id info are correct')
def step_impl(context):
    verify_payment_account_objects(context, context.response)
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.transferAmount - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress),
                           (context.destinationBalance + context.transferAmount))


@then('the balances after a payment with invalid invoice id info are correct')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "non-hexadecimal number found in fromhex() arg at position 3")
    verify_account_balance(get_account_balance(context, context.sourceAddress), context.sourceBalance)
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)


@then('the balances after a payment from preauth account are correct')
def step_impl(context):
    verify_payment_account_objects(context, context.response)
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.transferAmount - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress),
                           (context.destinationBalance + context.transferAmount))


@then('the balances after a payment from non-preauth account are correct')
def step_impl(context):
    verify_test_status(context.test_status, "failed")
    verify_test_error_message(str(context.exception), "Transaction failed: tecNO_PERMISSION")
    verify_account_balance(get_account_balance(context, context.sourceAddress),
                           (context.sourceBalance - context.minDrop))
    verify_account_balance(get_account_balance(context, context.destinationAddress), context.destinationBalance)
